New component - URL Preview List

ilmotta opened the pull request:
Fixes https://github.com/status-im/status-mobile/issues/15617

### Summary

This PR implements the the `URL Preview List` component. Please, refer to the [issue](https://github.com/status-im/status-mobile/issues/15617) for more details and the component specs.

### Review notes

- I have already tested this component in the chat composer and it works pretty well, so I think the implementation here is good for a review!
- I've inspected the component for needless re-renders, no issues here.
- The quo component `view` implements the pattern recently described by @ulisesmac. See his great explanation here https://github.com/status-im/status-mobile/issues/15552#issuecomment-1492590074.

### Demo

[demo.webm](https://user-images.githubusercontent.com/46027/231267304-decf8513-d813-4938-809c-e36c39a80c10.webm)

#### Platforms

- Android
- iOS

### Steps to test

Go to `Quo2.0 Preview` > `links` > `url-preview-list`

status: ready

ilmotta committed:
New component URL Preview List

ilmotta committed:
Auto jump to last element in the list

ilmotta committed:
Refactor: use better names

ilmotta committed:
Improve comment

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 8b42af40 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-15620/1/) | 2023-04-11 19:46:04 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-15620/1/consoleText) |

ulisesmac commented on src/quo2/components/links/url_preview_list/view.cljs:
```diff
@@ -0,0 +1,82 @@
+         :on-layout                         #(calculate-width preview-width horizontal-spacing %)
+         :horizontal                        true
+         :deceleration-rate                 :fast
+         :on-scroll-to-index-failed         identity
```
What is this `identity` used for? I tried commenting it out and it seems to behave the same

ilmotta replied:
Great question @ulisesmac. This might not be necessary, but I played safe here by copying what some other places do. See for instance this code comment screaming not to remove the workaround :laughing:  https://github.com/status-im/status-mobile/blob/develop/src/status_im2/contexts/chat/messages/list/view.cljs#L122 I can definitely remove it and if anything happens in the future we can fix it. Thanks for commenting on this.

In the Activity Center I once received feedback from QAs that after switching horizontally scrollable tabs, sometimes pressing on a tab would throw an exception. I never fully investigated why, but it certainly happened.

The documentation tells us this about `onScrollToIndexFailed`:

> Used to handle failures when scrolling to an index that has not been measured yet. Recommended action is to either compute your own offset and scrollTo it, or scroll as far as possible and then try again after more items have been rendered.

Now one thing is true, the `identity` function doesn't seem to be doing the recommended action at all. I wonder why the workaround even works... I'll remove this prop and see how it goes. Thanks!

ilmotta replied:
So @ulisesmac, I tried to change this and while experimenting with the component in another screen I got this error. It doesn't always happen, but I guess we'll need to keep the workaround.

<img src="https://user-images.githubusercontent.com/46027/231303479-131e16e6-1d89-4126-b64f-8aa96e9f8e5c.png" height="500" />

ulisesmac replied:
Ohh ok! :(  thanks for taking a look at that property!  Yes, let's keep that workaround

ulisesmac commented on src/quo2/components/links/url_preview_list/view.cljs:
```diff
@@ -0,0 +1,82 @@
+                       (fn []
+                         (when (and @flat-list-ref (pos? item-count))
+                           (.scrollToIndex ^js @flat-list-ref
+                                           #js
```
Is there a way to fix our linter to avoid adding that `\n` between the `#js` and `{`?

ilmotta replied:
I'd love that, I was puzzled by this formatting issue recently. There might be a way to configure something, or at the very least we can create an issue. The linter author is extremely helpful. I can try to take a look at this at the end of the week.

ilmotta replied:
Just found this from a week or so ago. So once 1.2.6 lands we will get a fix :)

https://github.com/kkinnear/zprint/issues/292#issuecomment-1496355633

ulisesmac replied:
Cool! 
thanks for sharing that comment

ulisesmac reviewed (approved):
Awesome work @ilmotta :tada: 
It feels and works very good :100:

ilmotta commented:
> Awesome work @ilmotta tada It feels and works very good 100

Thanks for the kind words and for reviewing so quickly @ulisesmac

smohamedjavid reviewed (approved):
Great work 🚀

briansztamfater reviewed (approved):
LGTM :)

ilmotta committed:
Use scrollToOffset to correctly jump to last element

Unfortunately FlatList.scrollToIndex can randomly fail, but scrollToOffset is
more reliable. I actually had no issues with scrollToOffset.

Also don't try to scroll if the item width is not yet calculated.

ilmotta committed:
Allow list items to be styled

ilmotta committed:
For now, remove the usage of resources namespace from spec
